### DISCUSSION TITLE
Error in 'fit_tune' and 'fit_chromaticity' Functions while running in computer cluster

### DISCUSSION THREAD

**Discussion by elafmusa** [atcollab/at#718] [Python AT]
I'm currently working with AT in a virtual environment within our DESY computer cluster, i am using Python version 3.9.7. However, I've encountered issues with the 'fit_tune' and 'fit_chromaticity' functions. These functions consistently break with a 'numpy.linalg.LinAlgErr: Singular matrix' error.

Despite using various lattices and seeds, I've faced the same issue. I've tried installing at both through pip and directly from the GitHub repository, yet the issue persists. 

I am attaching an example of the scripts i am running
[script.txt](https://github.com/atcollab/at/files/13550089/script.txt)
Noting the same scripts and functions works when i run it outside the cluster.

Any guidance or assistance you could offer on this matter would be greatly appreciated.

**Comment by swhite2401** [comment]
@simoneliuzzo recently proposed some changes that could be related to your issue, see #711

**Reply by elafmusa** [reply]
@swhite2401 Dear Simon, thank you for the information!

**Comment by simoneliuzzo** [comment]
Dear @elafmusa,

I had recently some issues with fit_tune and fit_chrom as well. For me the error was "unstable ring". So I proposed a pull request, and with the latest AT(master) you may set KStep=1e-9 for fit_tune and HStep=1e-8 for chromaticity. Those work for me for the FCCee@Z lattice optics.

An other thing that gave me problems are the FF FCC quads. You could check that the FF quadrupoles (Q[FD][01]... etc...) are not in the tune change loop. 

Also errors could be too big. 

Last, when running jobs in the cluster I specify the environment I want to use, to make sure I am running with the correct pyAT, pySC, etc... Also this gave me some issues. for slurm cluster in the sbatch launching script I simply add `source my_venv/bin/activate` before the python call.

I use python 3.8 at the moment. If none of the above solves your issue, I will try to run your script.

**Reply by elafmusa** [reply]
@simoneliuzzo Dear Simone,
Thank you for the very useful information, i have used the KStep and HStep in the fitting functions, and for FCC lattice i have exclude the FF quadrupoles for the tune fitting, My code works now!

Best Regards,
Elaf
